### PR TITLE
fix(runtime): handle lazy-instance promises for connected & disconnected callbacks

### PR DESCRIPTION
**Pull request by joewoodhouse**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [X] Build (`npm run build`) was run locally and any changes were pushed
- [X] Unit tests (`npm test`) were run locally and passed
- [X] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [X] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

Please check the type of change your PR introduces:
- [X] Bugfix
- [ ] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

See #4070

GitHub Issue Number: 4070

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

## Testing

Ran the unit tests (and added 2 new ones) and the karma tests
Also ran against my reproduction repository (see #4070) which fixed the issue

### CODE REVIEW

File: src/runtime/connected-callback.ts [hunk 1]
```diff
@@ -106,7 +106,11 @@ export const connectedCallback = (elm: d.HostElement) => {
       addHostEventListeners(elm, hostRef, cmpMeta.$listeners$, false);
 
       // fire off connectedCallback() on component instance
-      fireConnectedCallback(hostRef.$lazyInstance$);
+      if (hostRef.$lazyInstance$) {
+        fireConnectedCallback(hostRef.$lazyInstance$);
+      } else if (hostRef.$onReadyPromise$) {
+        hostRef.$onReadyPromise$.then(() => fireConnectedCallback(hostRef.$lazyInstance$));
+      }
```
Comment (tanner-reits) [inline src/runtime/connected-callback.ts:113]: Just a small change to resolve a SNC:
```suggestion
      if (hostRef?.$lazyInstance$) {
        fireConnectedCallback(hostRef.$lazyInstance$);
      } else if (hostRef?.$onReadyPromise$) {
        hostRef.$onReadyPromise$.then(() => fireConnectedCallback(hostRef.$lazyInstance$));
      }
```

File: src/runtime/disconnected-callback.ts [hunk 2]
```diff
@@ -17,11 +25,12 @@ export const disconnectedCallback = (elm: d.HostElement) => {
...
-      safeCall(instance, 'disconnectedCallback');
-    }
-    if (BUILD.cmpDidUnload) {
-      safeCall(instance, 'componentDidUnload');
+    if (!BUILD.lazyLoad) {
+      disconnectInstance(elm);
+    } else if (hostRef.$lazyInstance$) {
+      disconnectInstance(hostRef.$lazyInstance$);
+    } else if (hostRef.$onReadyPromise$) {
+      hostRef.$onReadyPromise$.then(() => disconnectInstance(hostRef.$lazyInstance$));
```
Comment (tanner-reits) [inline src/runtime/disconnected-callback.ts:33]: Just a small change to resolve a SNC:
```suggestion
    } else if (hostRef?.$lazyInstance$) {
      disconnectInstance(hostRef.$lazyInstance$);
    } else if (hostRef?.$onReadyPromise$) {
      hostRef.$onReadyPromise$.then(() => disconnectInstance(hostRef.$lazyInstance$));
```

**Review by tanner-reits — COMMENTED** [review]
Sorry for the (very) delayed response here, @joewoodhouse. The team finally got around to checking out the associated issue and this work. I tested the changes using the `dist` and `dist-custom-elements` targets, as well as with SSR (just to be safe). Everything looks good!

I resolved the merge conflicts for the time being. And I believe my only ask to get this merged back into Stencil would be to resolve some of the new strict null checks (SNCs). I made note of those in the review to make things easy. Once that is all squared-away, I'll work with the team to get this merged and released! 

Thanks again!

**Comment by joewoodhouse** [comment]
Hey @tanner-reits just looked at those and merged them in - thanks for the suggestions, hadn't spotted the new SNC stuff. Let me know if you need anything else 👍

**Review by tanner-reits — APPROVED** [review]

**Comment by danyball** [comment]
What exactly is this PR fixing @joewoodhouse ? Some of our Angular apps using stencil components have issues now. In my case the parent disconnectedCallback is running, but no disconnection at the children.

**Comment by joewoodhouse** [comment]
@danyball fixing this bug https://github.com/ionic-team/stencil/issues/4070 as it says

**Comment by danyball** [comment]
Just found the link, thanks. Do you use your components in an angular environment?

**Comment by danyball** [comment]
@joewoodhouse Because you are deep in this logic: Is it by design that such an order can occur now?

The components:
```
<select>
  <option>
  <option>
</select>
```

Order:
```
option connect 
option connect 
select connect
option disconnect
option connect 
option disconnect
option connect 
select disconnect --> without option disconnect? There are childs. Or could it happen because its called async?
select connect --> without option disconnect? There are childs.
```

**Comment by joewoodhouse** [comment]
@danyball can you create a minimal code repo that demonstrates the problem? Doesn't look correct to me

**Comment by danyball** [comment]
Thats my problem, I cant. It just happens in an angular app with many of ngIf, ng-template, formControl. I just debugged, that the script is going into `else if (hostRef?.$onReadyPromise$)` in that angular app, and in my simple angular app it always runs into `else if (hostRef?.$lazyInstance$)`. 

Maybe you could tell me the difference? Or in which case `$onReadyPromise$` is set? 
I also saw that the "hydrated" class is not set to select and option component when disconnection happens. Maybe you could imagine in which situation this could happen?

Sorry for all these theoretical questions, I really try to create a reproducer for that.

**Pull request merged**